piop/src/sumcheck/tests/utils.rs

```diff
@@ -0,0 +1,110 @@
+use ark_std::{end_timer, start_timer, vec::Vec};
+use crypto_bigint::Random;
+use crypto_primitives::PrimeField;
+use rand::{Rng, RngCore};
+use std::sync::Arc;
+use zinc_poly::{mle::DenseMultilinearExtension, utils::ArithErrors};
+
+#[allow(clippy::arithmetic_side_effects)]
+#[allow(clippy::type_complexity)]
+pub(crate) fn rand_poly<F: PrimeField + Random, Rn: RngCore>(
+    nv: usize,
+    num_multiplicands_range: (usize, usize),
+    num_products: usize,
+    rng: &mut Rn,
+    config: &F::Config,
+) -> Result<
+    (
+        (Vec<DenseMultilinearExtension<F>>, usize),
+        Vec<(F, Vec<usize>)>,
+        F,
+    ),
+    ArithErrors,
+> {
+    let mut sum = F::zero_with_cfg(config);
+    let mut mles = vec![];
+    let mut products = Vec::with_capacity(num_products);
+    let mut degree = 0;
+    let mut current_mle_index = 0;
+    for _ in 0..num_products {
+        let num_multiplicands =
+            rng.random_range(num_multiplicands_range.0..num_multiplicands_range.1);
+        degree = num_multiplicands.max(degree);
+        let (product, product_sum) = random_mle_list(nv, num_multiplicands, rng, config.clone());
+        let product = product
+            .into_iter()
+            .map(|p| Arc::into_inner(p).unwrap())
+            .collect::<Vec<_>>();
+
+        let coefficient = F::try_random(rng).expect("sampling coefficient failed");
+        mles.extend(product);
+        sum += &(product_sum * &coefficient);
+
+        let indices: Vec<usize> =
+            (current_mle_index..current_mle_index + num_multiplicands).collect();
+        products.push((coefficient, indices));
+        current_mle_index += num_multiplicands;
+    }
+
+    Ok(((mles, degree), products, sum))
+}
+
+pub fn rand_poly_comb_fn<F: PrimeField>(
+    vals: &[F],
+    products: &[(F, Vec<usize>)],
+    config: F::Config,
+) -> F {
+    let mut result = F::zero_with_cfg(&config);
+    for (coef, indices) in products {
+        let mut term = coef.clone();
+        for &i in indices {
+            term *= &vals[i];
+        }
+        result += &term;
+    }
+
+    result
+}
+
+/// Sample a random list of multilinear polynomials.
+/// Returns
+/// - the list of polynomials,
+/// - its sum of polynomial evaluations over the boolean hypercube.
+pub fn random_mle_list<F: PrimeField + Random, Rn: RngCore>(
+    nv: usize,
+    degree: usize,
+    rng: &mut Rn,
+    config: F::Config,
+) -> (Vec<Arc<DenseMultilinearExtension<F>>>, F) {
+    let start = start_timer!(|| "sample random mle list");
+    let mut multiplicands = Vec::with_capacity(degree);
+    for _ in 0..degree {
+        multiplicands.push(Vec::with_capacity(1 << nv));
+    }
```

Comment:
In general it's more preferrable to use map/collect one-liners, but in this case there's a convenient macro:
```suggestion
    let mut multiplicands = vec![Vec::with_capacity(1 << nv); degree];
```

Reply:
Using vec! macro here isn't great as it clones under the hood and `Vec::clone` doesn't respect capacity.
https://stackoverflow.com/questions/77136991/why-does-vecvecwith-capacityn-create-0-capacity-for-the-sub-vectors

Reply:
Interesting. In this case I would go with `(0..degree).map(|_| Vec::with_capacity(1 << nv))).collect_vec()`

Reply:
Not that it matters much as it's in tests. If it's ever to be copied into non-test parts of the code let's simplify the job for whoever does that :)